File: scripts/export-agent-chats.ts
#!/usr/bin/env ts-node

import * as fs from 'fs';
import * as path from 'path';
import Database from 'better-sqlite3';

const DB_PATH = path.join(
  process.env.HOME!,
  'Library/Application Support/Cursor/User/globalStorage/state.vscdb'
);

const OUTPUT_DIR = path.join(__dirname, '../docs/cursor-agent-history');

interface BubbleData {
  bubbleId?: string;
  messages?: Array<{
    type?: string;
    text?: string;
    role?: string;
    content?: string;
    timestamp?: number;
  }>;
  [key: string]: any;
}

function sanitizeFilename(str: string): string {
  return str.replace(/[^a-z0-9]/gi, '_').substring(0, 50);
}

function extractChats() {
  console.log('Opening database:', DB_PATH);
  
  const db = new Database(DB_PATH, { readonly: true });
  
  // Get all bubble IDs
  const bubbles = db.prepare(
    "SELECT key, value FROM cursorDiskKV WHERE key LIKE 'bubbleId:%'"
  ).all();
  
  console.log(`Found ${bubbles.length} agent conversations`);
  
  let exported = 0;
  const errors: string[] = [];
  
  for (const bubble of bubbles) {
    try {
      const data: BubbleData = JSON.parse(bubble.value as string);
      const bubbleId = bubble.key.split(':').slice(2).join(':') || data.bubbleId || 'unknown';
      
      // Try to extract meaningful content
      let markdown = `# Agent Conversation\n\n`;
      markdown += `**Bubble ID:** ${bubbleId}\n`;
      markdown += `**Key:** ${bubble.key}\n\n`;
      markdown += `---\n\n`;
      
      // Extract messages if they exist
      if (data.messages && Array.isArray(data.messages)) {
        for (const msg of data.messages) {
          const role = msg.role || msg.type || 'unknown';
          const content = msg.text || msg.content || '';
          const timestamp = msg.timestamp ? new Date(msg.timestamp).toISOString() : '';
          
          if (content) {
            markdown += `## ${role.toUpperCase()}${timestamp ? ` (${timestamp})` : ''}\n\n`;
            markdown += `${content}\n\n`;
            markdown += `---\n\n`;
          }
        }
      }
      
      // If no messages found, dump the raw data
      if (markdown.split('---').length <= 3) {
        markdown += `## Raw Data\n\n`;
        markdown += '```json\n';
        markdown += JSON.stringify(data, null, 2);
        markdown += '\n```\n';
      }
      
      const filename = `${sanitizeFilename(bubbleId)}_${Date.now()}.md`;
      const filepath = path.join(OUTPUT_DIR, filename);
      
      fs.writeFileSync(filepath, markdown);
      exported++;
      
      if (exported % 100 === 0) {
        console.log(`Exported ${exported}/${bubbles.length}...`);
      }
    } catch (err) {
      errors.push(`Error processing ${bubble.key}: ${err}`);
    }
  }
  
  db.close();
  
  console.log(`\n✅ Exported ${exported} conversations to ${OUTPUT_DIR}`);
  
  if (errors.length > 0) {
    console.log(`\n⚠️  ${errors.length} errors occurred:`);
    errors.slice(0, 10).forEach(e => console.log(`  - ${e}`));
  }
  
  // Create index file
  const indexContent = `# Cursor Agent Chat History Export\n\n` +
    `**Exported:** ${new Date().toISOString()}\n` +
    `**Total Conversations:** ${exported}\n\n` +
    `All your agent conversations have been exported to individual markdown files in this directory.\n\n` +
    `## Notes\n\n` +
    `- Each file represents one agent conversation\n` +
    `- Files are named by their bubble ID\n` +
    `- This is a backup of your chat history from the old workspace\n`;
  
  fs.writeFileSync(path.join(OUTPUT_DIR, 'README.md'), indexContent);
  console.log(`\n📝 Created index file: README.md`);
}

try {
  extractChats();
} catch (err) {
  console.error('Fatal error:', err);
  process.exit(1);
}

File: scripts/export-agent-chats.py
#!/usr/bin/env python3

import sqlite3
import json
import os
from pathlib import Path
from datetime import datetime
import re

# Paths
HOME = os.path.expanduser("~")
DB_PATH = os.path.join(HOME, "Library/Application Support/Cursor/User/globalStorage/state.vscdb")
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), "../docs/cursor-agent-history")

def sanitize_filename(text, max_length=50):
    """Create a safe filename from text"""
    text = re.sub(r'[^a-zA-Z0-9_-]', '_', text)
    return text[:max_length]

def extract_text_content(data, max_depth=3, current_depth=0):
    """Recursively extract text content from nested data structures"""
    texts = []
    
    if current_depth > max_depth:
        return texts
    
    if isinstance(data, dict):
        for key, value in data.items():
            if key in ['text', 'content', 'message', 'query', 'response']:
                if isinstance(value, str) and len(value) > 10:
                    texts.append((key, value))
            else:
                texts.extend(extract_text_content(value, max_depth, current_depth + 1))
    elif isinstance(data, list):
        for item in data:
            texts.extend(extract_text_content(item, max_depth, current_depth + 1))
    
    return texts

def export_chats():
    print(f"Opening database: {DB_PATH}")
    
    if not os.path.exists(DB_PATH):
        print(f"❌ Database not found at {DB_PATH}")
        return
    
    # Create output directory
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    
    # Connect to database
    conn = sqlite3.connect(DB_PATH)
    cursor = conn.cursor()
    
    # Get all bubble IDs
    cursor.execute("SELECT key, value FROM cursorDiskKV WHERE key LIKE 'bubbleId:%'")
    bubbles = cursor.fetchall()
    
    print(f"Found {len(bubbles)} agent conversations")
    
    exported = 0
    errors = []
    
    for key, value_blob in bubbles:
        try:
            # Parse JSON data
            data = json.loads(value_blob)
            
            # Extract bubble ID from key
            parts = key.split(':')
            bubble_id = parts[-1] if len(parts) > 2 else 'unknown'
            
            # Start markdown content
            markdown = f"# Agent Conversation\n\n"
            markdown += f"**Bubble ID:** {bubble_id}\n"
            markdown += f"**Full Key:** {key}\n"
            markdown += f"**Exported:** {datetime.now().isoformat()}\n\n"
            markdown += "---\n\n"
            
            # Extract any text content
            text_contents = extract_text_content(data)
            
            if text_contents:
                markdown += "## Conversation Content\n\n"
                for label, content in text_contents:
                    markdown += f"### {label.title()}\n\n"
                    markdown += f"{content}\n\n"
                    markdown += "---\n\n"
            
            # Add raw data section
            markdown += "## Raw Data Structure\n\n"
            markdown += "```json\n"
            markdown += json.dumps(data, indent=2, ensure_ascii=False)
            markdown += "\n```\n"
            
            # Save file
            filename = f"{sanitize_filename(bubble_id)}_{exported:04d}.md"
            filepath = os.path.join(OUTPUT_DIR, filename)
            
            with open(filepath, 'w', encoding='utf-8') as f:
                f.write(markdown)
            
            exported += 1
            
            if exported % 100 == 0:
                print(f"Exported {exported}/{len(bubbles)}...")
        
        except Exception as err:
            errors.append(f"Error processing {key}: {str(err)}")
    
    conn.close()
    
    print(f"\n✅ Exported {exported} conversations to {OUTPUT_DIR}")
    
    if errors:
        print(f"\n⚠️  {len(errors)} errors occurred:")
        for error in errors[:10]:
            print(f"  - {error}")
    
    # Create index file
    index_content = f"""# Cursor Agent Chat History Export

**Exported:** {datetime.now().isoformat()}
**Total Conversations:** {exported}
**Database:** {DB_PATH}

All your agent conversations have been exported to individual markdown files in this directory.

## Files

Each file is named with the format: `<bubble_id>_<number>.md`

## Notes

- Each file represents one agent conversation (bubble)
- Files contain both extracted content and raw JSON data
- This is a complete backup of your chat history from Cursor's global storage
- You can search these files with any text editor or grep

## How to Use

1. Use your IDE's search feature to find specific conversations
2. Each file has a unique bubble ID for reference
3. The raw JSON data is included for complete context
"""
    
    readme_path = os.path.join(OUTPUT_DIR, "README.md")
    with open(readme_path, 'w') as f:
        f.write(index_content)
    
    print(f"\n📝 Created index file: {readme_path}")
    print(f"\n🎉 All done! Check {OUTPUT_DIR}")

if __name__ == "__main__":
    try:
        export_chats()
    except Exception as err:
        print(f"❌ Fatal error: {err}")
        import traceback
        traceback.print_exc()
        exit(1)


